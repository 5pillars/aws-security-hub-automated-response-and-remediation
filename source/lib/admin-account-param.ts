// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';

export default class AdminAccountParam extends Construct {
  public readonly value: string;

  constructor(scope: Construct, id: string) {
    super(scope, id);

    const validAwsAccount = '^\\d{12}$';

    const param = new cdk.CfnParameter(this, 'Admin Account Number', {
      description: 'Admin account number',
      type: 'String',
      allowedPattern: validAwsAccount,
    });
    param.overrideLogicalId(`SecHubAdminAccount`);
    this.value = param.valueAsString;
  }
}
