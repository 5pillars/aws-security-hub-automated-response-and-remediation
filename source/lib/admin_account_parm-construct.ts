// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';

export class AdminAccountParm extends Construct {
  public readonly adminAccountNumber: cdk.CfnParameter;

  constructor(scope: Construct, id: string) {
    super(scope, id);

    const validAwsAccount = '\\d{12}';

    this.adminAccountNumber = new cdk.CfnParameter(this, 'Admin Account Number', {
      description: 'Admin account number',
      type: 'String',
      allowedPattern: validAwsAccount,
    });
    this.adminAccountNumber.overrideLogicalId(`SecHubAdminAccount`);
  }
}
