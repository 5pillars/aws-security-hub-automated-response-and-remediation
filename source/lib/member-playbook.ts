// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { CfnCondition, CfnParameter, CfnResource, Fn, Stack } from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { SerializedNestedStackFactory } from './cdk-helper/nested-stack';

export interface MemberPlaybookProps {
  readonly name: string;
  readonly nestedStackFactory: SerializedNestedStackFactory;
  readonly parameters?: { [_: string]: string };
  readonly stackDependencies?: CfnResource[];
  readonly defaultState?: 'yes' | 'no';
  readonly description?: string;
}

export class MemberPlaybookNew {
  parameterName = '';
  playbookStacks: Stack[];

  constructor(scope: Construct, props: MemberPlaybookProps) {
    const templateFileOriginal = `${props.name}MemberStackOriginal.template`;
    const templateFileNew = `${props.name}MemberStackNew.template`;
    const illegalChars = /[\\._]/g;
    const playbookName = props.name.replace(illegalChars, '');
    this.parameterName = `Load${playbookName}MemberStacks`;

    //---------------------------------------------------------------------
    // Playbook Template Nested Stack
    //
    const stackOptionOriginal = new CfnParameter(scope, `LoadOriginalMemberStack${playbookName}`, {
      type: 'String',
      description:
        props.description ?? `Install the member components for automated remediation of ${props.name} controls?`,
      default: props.defaultState ?? 'no',
      allowedValues: ['yes', 'no'],
    });
    const stackOptionNew = new CfnParameter(scope, `LoadNewMemberStack${playbookName}`, {
      type: 'String',
      description:
        props.description ?? `Install the member components for automated remediation of ${props.name} controls?`,
      default: props.defaultState ?? 'no',
      allowedValues: ['yes', 'no'],
    });
    stackOptionOriginal.overrideLogicalId(this.parameterName);
      this.playbookStacks = []
      const originalNestedStack = props.nestedStackFactory.addNestedStack(`PlaybookMemberStackOriginal${playbookName}`, {
        templateRelativePath: `playbooks/${templateFileOriginal}`,
        parameters: props.parameters,
        condition: new CfnCondition(scope, `load${playbookName}CondOriginal`, {
          expression: Fn.conditionEquals(stackOptionOriginal, 'yes'),
        }),
      });

    const cfnStackOriginal = originalNestedStack.nestedStackResource as CfnResource;
    cfnStackOriginal.overrideLogicalId(`PlaybookMemberStackOriginal${props.name}`);

    const newNestedStack = props.nestedStackFactory.addNestedStack(`PlaybookMemberStackNew${playbookName}`, {
      templateRelativePath: `playbooks/${templateFileNew}`,
      parameters: props.parameters,
      condition: new CfnCondition(scope, `load${playbookName}CondNew`, {
        expression: Fn.conditionEquals(stackOptionNew, 'yes'),
      }),
    });

    const cfnStackNew = newNestedStack.nestedStackResource as CfnResource;
    cfnStackNew.overrideLogicalId(`PlaybookMemberStackNew${props.name}`);

    this.playbookStacks.push(originalNestedStack as Stack)
    this.playbookStacks.push(newNestedStack as Stack)
  }
}

export class MemberPlaybook {
  parameterName = '';
  playbookStack: Stack;

  constructor(scope: Construct, props: MemberPlaybookProps) {
    const templateFile = `${props.name}MemberStack.template`;
    const illegalChars = /[\\._]/g;
    const playbookName = props.name.replace(illegalChars, '');
    this.parameterName = `Load${playbookName}MemberStack`;

    //---------------------------------------------------------------------
    // Playbook Template Nested Stack
    //
    const stackOption = new CfnParameter(scope, `LoadMemberStack${playbookName}`, {
      type: 'String',
      description:
        props.description ?? `Install the member components for automated remediation of ${props.name} controls?`,
      default: props.defaultState ?? 'no',
      allowedValues: ['yes', 'no'],
    });
    stackOption.overrideLogicalId(this.parameterName);
    this.playbookStack = props.nestedStackFactory.addNestedStack(`PlaybookMemberStack${playbookName}`, {
      templateRelativePath: `playbooks/${templateFile}`,
      parameters: props.parameters,
      condition: new CfnCondition(scope, `load${playbookName}Cond`, {
        expression: Fn.conditionEquals(stackOption, 'yes'),
      }),
    });
    const cfnStack = this.playbookStack.nestedStackResource as CfnResource;
    cfnStack.overrideLogicalId(`PlaybookMemberStack${props.name}`);
  }
}
