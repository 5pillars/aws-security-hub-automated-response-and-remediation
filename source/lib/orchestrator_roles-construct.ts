#!/usr/bin/env node
/*****************************************************************************
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.   *
 *                                                                            *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may   *
 *  not use this file except in compliance with the License. A copy of the    *
 *  License is located at                                                     *
 *                                                                            *
 *      http://www.apache.org/licenses/LICENSE-2.0                            *
 *                                                                            *
 *  or in the 'license' file accompanying this file. This file is distributed *
 *  on an 'AS IS' BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,        *
 *  express or implied. See the License for the specific language governing   *
 *  permissions and limitations under the License.                            *
 *****************************************************************************/

import * as cdk from '@aws-cdk/core';
import { 
    PolicyStatement, 
    Effect, 
    Role, 
    PolicyDocument, 
    ArnPrincipal,
    ServicePrincipal,
    CompositePrincipal, 
    CfnRole 
} from '@aws-cdk/aws-iam';

export interface OrchRoleProps {
    solutionId: string;
    adminAccountId: string;
    adminRoleName: string;
}

export class OrchestratorMemberRole extends cdk.Construct {
  constructor(scope: cdk.Construct, id: string, props: OrchRoleProps) {
    super(scope, id);
    const RESOURCE_PREFIX = props.solutionId.replace(/^DEV-/,''); // prefix on every resource name
    const stack = cdk.Stack.of(this);
    const memberPolicy = new PolicyDocument();

    /**
     * @description Cross-account permissions for Orchestration role
     * @type {PolicyStatement}
     */
    const iamPerms = new PolicyStatement();
    iamPerms.addActions(
        "iam:PassRole",
        "iam:GetRole"
    )
    iamPerms.effect = Effect.ALLOW
    iamPerms.addResources(
        `arn:${stack.partition}:iam::${stack.account}:role/${RESOURCE_PREFIX}-*`
    );
    memberPolicy.addStatements(iamPerms)
    const ssmROPerms = new PolicyStatement()
    ssmROPerms.addActions(
        "ssm:DescribeAutomationExecutions",
        "ssm:DescribeDocument",
        "ssm:GetParameters"
    )
    ssmROPerms.effect = Effect.ALLOW;
    ssmROPerms.addResources(
        "arn:" + stack.partition + ":ssm:*:*:*"
    )
    memberPolicy.addStatements(ssmROPerms)

    const ssmRWPerms = new PolicyStatement()
    ssmRWPerms.addActions(
        "ssm:StartAutomationExecution",
        "ssm:GetAutomationExecution"
    )
    ssmRWPerms.addResources(
        // `arn:${stack.partition}:ssm:*:${stack.account}:document/SHARR-*`,
        // `arn:${stack.partition}:ssm:*:${stack.account}:automation-definition/*`,
        // `arn:${stack.partition}:ssm:*:${stack.account}:document/SHARR-*`,
        stack.formatArn({
            service: 'ssm',
            region: '*',
            resource: 'document',
            resourceName: 'SHARR-*',
            sep: '/'
        }),
        stack.formatArn({
            service: 'ssm',
            region: '*',
            resource: 'automation-definition',
            resourceName: '*',
            sep: '/'
        }),
        stack.formatArn({
            service: 'ssm',
            region: '*',
            resource: 'automation-definition',
            account:'',
            resourceName: '*',
            sep: '/'
        }),
        stack.formatArn({
            service: 'ssm',
            region: '*',
            resource: 'automation-execution',
            resourceName: '*',
            sep: '/'
        })
    );
    memberPolicy.addStatements(ssmRWPerms)

    const sechubPerms = new PolicyStatement();
    sechubPerms.addActions("cloudwatch:PutMetricData")
    sechubPerms.addActions("securityhub:BatchUpdateFindings")
    sechubPerms.effect = Effect.ALLOW
    sechubPerms.addResources("*")

    memberPolicy.addStatements(sechubPerms)

    let principalPolicyStatement = new PolicyStatement();

    principalPolicyStatement.addActions("sts:AssumeRole");
    principalPolicyStatement.effect = Effect.ALLOW;

    let roleprincipal = new ArnPrincipal(
        `arn:${stack.partition}:iam::${props.adminAccountId}:role/${props.adminRoleName}`
    );

    let principals = new CompositePrincipal(roleprincipal);
    principals.addToPolicy(principalPolicyStatement);

    let serviceprincipal = new ServicePrincipal('ssm.amazonaws.com')
    principals.addPrincipals(serviceprincipal);

    let memberRole = new Role(this, 'MemberAccountRole', {
        assumedBy: principals,
        inlinePolicies: {
            'member_orchestrator': memberPolicy
        },
        roleName: `${RESOURCE_PREFIX}-SHARR-Orchestrator-Member`
    });

    const memberRoleResource = memberRole.node.findChild('Resource') as CfnRole;

    memberRoleResource.cfnOptions.metadata = {
        cfn_nag: {
            rules_to_suppress: [{
                id: 'W11',
                reason: 'Resource * is required due to the administrative nature of the solution.'
            },{
                id: 'W28',
                reason: 'Static names chosen intentionally to provide integration in cross-account permissions'
            }]
        }
    }
  }
}