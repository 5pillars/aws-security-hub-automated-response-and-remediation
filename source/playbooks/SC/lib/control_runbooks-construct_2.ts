// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
import { Runtime } from 'aws-cdk-lib/aws-lambda';
import { Construct } from 'constructs';
import { ControlRunbookDocument } from '../ssmdocs/control_runbook';
import { CfnCondition, CfnParameter, Fn } from 'aws-cdk-lib';

import * as cloudfront_1 from '../ssmdocs/SC_CloudFront.1';
import * as cloudfront_12 from '../ssmdocs/SC_CloudFront.12';
import * as codebuild_5 from '../ssmdocs/SC_CodeBuild.5';
import * as ec2_4 from '../ssmdocs/SC_EC2.4';
import * as ec2_8 from '../ssmdocs/SC_EC2.8';
import * as ec2_18 from '../ssmdocs/SC_EC2.18';
import * as ec2_19 from '../ssmdocs/SC_EC2.19';
import * as ec2_23 from '../ssmdocs/SC_EC2.23';
import * as ecr_1 from '../ssmdocs/SC_ECR.1';
import * as s3_11 from '../ssmdocs/SC_S3.11';
import * as s3_13 from '../ssmdocs/SC_S3.13';
import * as secretsmanager_1 from '../ssmdocs/SC_SecretsManager.1';
import * as secretsmanager_3 from '../ssmdocs/SC_SecretsManager.3';
import * as secretsmanager_4 from '../ssmdocs/SC_SecretsManager.4';
import * as ssm_4 from '../ssmdocs/SC_SSM.4';
import * as apigateway_3 from '../ssmdocs/SC_APIGateway.3';
import * as cloudfront_5 from '../ssmdocs/SC_CloudFront.5';
import * as cloudfront_2 from '../ssmdocs/SC_CloudFront.2';
import * as documentdb_2 from '../ssmdocs/SC_DocumentDB.2';
import * as dynamodb_2 from '../ssmdocs/SC_DynamoDB.2';
import * as dynamodb_1 from '../ssmdocs/SC_DynamoDB.1';
import * as eks_8 from '../ssmdocs/SC_EKS.8';
import * as elb_4 from '../ssmdocs/SC_ELB.4';
import * as elb_6 from '../ssmdocs/SC_ELB.6';
import * as neptune_2 from '../ssmdocs/SC_Neptune.2';
import * as neptune_5 from '../ssmdocs/SC_Neptune.5';
import * as neptune_4 from '../ssmdocs/SC_Neptune.4';
import * as rds_17 from '../ssmdocs/SC_RDS.17';
import * as rds_11 from '../ssmdocs/SC_RDS.11';
import * as s3_14 from '../ssmdocs/SC_S3.14';
import * as stepfunctions_1 from '../ssmdocs/SC_StepFunctions.1';
import * as waf_11 from '../ssmdocs/SC_WAF.11';

export interface PlaybookProps {
  standardShortName: string;
  standardLongName: string;
  standardVersion: string;
  runtimePython: Runtime;
  solutionId: string;
  solutionAcronym: string;
  solutionVersion: string;
}

export class ControlRunbooks extends Construct {
  protected readonly standardLongName: string;
  protected readonly standardVersion: string;
  protected controls: Set<string> = new Set<string>();

  constructor(scope: Construct, id: string, props: PlaybookProps) {
    super(scope, id);

    this.standardLongName = props.standardLongName;
    this.standardVersion = props.standardVersion;
    this.add(apigateway_3.createControlRunbook(this, 'APIGateway.3', props));
    this.add(cloudfront_1.createControlRunbook(this, 'CloudFront.1', props));
    this.add(cloudfront_2.createControlRunbook(this, 'CloudFront.2', props));
    this.add(cloudfront_5.createControlRunbook(this, 'CloudFront.5', props));
    this.add(cloudfront_12.createControlRunbook(this, 'CloudFront.12', props));
    this.add(codebuild_5.createControlRunbook(this, 'CodeBuild.5', props));
    this.add(documentdb_2.createControlRunbook(this, 'DocumentDB.2', props));
    this.add(dynamodb_1.createControlRunbook(this, 'DynamoDB.1', props));
    this.add(dynamodb_2.createControlRunbook(this, 'DynamoDB.2', props));
    this.add(ec2_4.createControlRunbook(this, 'EC2.4', props));
    this.add(ec2_8.createControlRunbook(this, 'EC2.8', props));
    this.add(ec2_18.createControlRunbook(this, 'EC2.18', props));
    this.add(ec2_19.createControlRunbook(this, 'EC2.19', props));
    this.add(ec2_23.createControlRunbook(this, 'EC2.23', props));
    this.add(ecr_1.createControlRunbook(this, 'ECR.1', props));
    this.add(eks_8.createControlRunbook(this, 'EKS.8', props));
    this.add(elb_4.createControlRunbook(this, 'ELB.4', props));
    this.add(elb_6.createControlRunbook(this, 'ELB.6', props));
    this.add(neptune_2.createControlRunbook(this, 'Neptune.2', props));
    this.add(neptune_4.createControlRunbook(this, 'Neptune.4', props));
    this.add(neptune_5.createControlRunbook(this, 'Neptune.5', props));
    this.add(rds_11.createControlRunbook(this, 'RDS.11', props));
    this.add(rds_17.createControlRunbook(this, 'RDS.17', props));
    this.add(s3_11.createControlRunbook(this, 'S3.11', props));
    this.add(s3_13.createControlRunbook(this, 'S3.13', props));
    this.add(s3_14.createControlRunbook(this, 'S3.14', props));
    this.add(secretsmanager_1.createControlRunbook(this, 'SecretsManager.1', props));
    this.add(secretsmanager_3.createControlRunbook(this, 'SecretsManager.3', props));
    this.add(secretsmanager_4.createControlRunbook(this, 'SecretsManager.4', props));
    this.add(ssm_4.createControlRunbook(this, 'SSM.4', props));
    this.add(stepfunctions_1.createControlRunbook(this, 'StepFunctions.1', props));
    this.add(waf_11.createControlRunbook(this, 'WAF.11', props));
  }

  protected add(document: ControlRunbookDocument) {
    const controlId = document.getControlId();
    const enableParamDescription = this.getEnableParamDescription(controlId);
    const enableParamValueAvailable = 'Available';
    const enableParam = new CfnParameter(this, `Enable ${controlId}`, {
      type: 'String',
      description: enableParamDescription,
      default: enableParamValueAvailable,
      allowedValues: [enableParamValueAvailable, 'NOT Available'],
    });

    const installSsmDoc = new CfnCondition(this, `Enable ${controlId} Condition`, {
      expression: Fn.conditionEquals(enableParam, enableParamValueAvailable),
    });

    document.cfnDocument.cfnOptions.condition = installSsmDoc;

    this.controls.add(document.getControlId());
  }

  protected getEnableParamDescription(controlId: string) {
    // eslint-disable-next-line prettier/prettier
    return (
      `Enable/disable availability of remediation for ${this.standardLongName} version ` +
      `${this.standardVersion} Control ${controlId} in Security Hub Console Custom Actions. If ` +
      'NOT Available the remediation cannot be triggered from the Security Hub console in the ' +
      'Security Hub Admin account.'
    );
  }

  public has(controlId: string): boolean {
    return this.controls.has(controlId);
  }
}
