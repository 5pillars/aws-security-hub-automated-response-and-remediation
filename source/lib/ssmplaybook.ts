#!/usr/bin/env node
/*****************************************************************************
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.   *
 *                                                                            *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may   *
 *  not use this file except in compliance with the License. A copy of the    *
 *  License is located at                                                     *
 *                                                                            *
 *      http://www.apache.org/licenses/LICENSE-2.0                            *
 *                                                                            *
 *  or in the 'license' file accompanying this file. This file is distributed *
 *  on an 'AS IS' BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,        *
 *  express or implied. See the License for the specific language governing   *
 *  permissions and limitations under the License.                            *
 *****************************************************************************/

import * as cdk from '@aws-cdk/core';
import * as ssm from '@aws-cdk/aws-ssm';
import * as fs from 'fs';
import * as yaml from 'js-yaml';
import * as sfn from '@aws-cdk/aws-stepfunctions';
import * as events from '@aws-cdk/aws-events';
import {
    Effect, 
    PolicyStatement, 
    ServicePrincipal, 
    Policy, 
    Role, 
    CfnRole,
    ArnPrincipal,
    CompositePrincipal
} from '@aws-cdk/aws-iam';
import { StateMachine } from '@aws-cdk/aws-stepfunctions';
import { IRuleTarget, EventPattern, Rule } from '@aws-cdk/aws-events';
import { MemberRoleStack } from '../solution_deploy/lib/remediation_runbook-stack';

/*
 * @author AWS Solutions Development
 * @description SSM-based remediation parameters
 * @type {playbookConstruct}
 */

export interface IssmPlaybookProps {
    securityStandard: string;     // ex. AFSBP
    securityStandardVersion: string;
    controlId: string;
    ssmDocPath: string;
    ssmDocFileName: string;
    solutionVersion: string;
    solutionDistBucket: string;
    adminRoleName?: string;
    remediationPolicy?: Policy;
    adminAccountNumber?: string;
    solutionId?: string;
    scriptPath?: string;
}

export class SsmPlaybook extends cdk.Construct {
    
  constructor(scope: cdk.Construct, id: string, props: IssmPlaybookProps) {
    super(scope, id);
    
    let scriptPath = ''
    if (props.scriptPath == undefined ) {
        scriptPath = `${props.ssmDocPath}/scripts`
    } else {
        scriptPath = props.scriptPath
    }

    let illegalChars = /[\.]/g;

    const enableParam = new cdk.CfnParameter(this, 'Enable ' + props.controlId, {
        type: "String",
        description: `Enable/disable availability of remediation for ${props.securityStandard} version ${props.securityStandardVersion} Control ${props.controlId} in Security Hub Console Custom Actions. If NOT Available the remediation cannot be triggered from the Security Hub console in the Security Hub Admin account.`,
        default: "Available",
        allowedValues: ["Available", "NOT Available"]
    })
    enableParam.overrideLogicalId(`${props.securityStandard}${props.controlId.replace(illegalChars, '')}Active`)

    const installSsmDoc = new cdk.CfnCondition(this, 'Enable ' + props.controlId + ' Condition', {
        expression: cdk.Fn.conditionEquals(enableParam, "Available")
    })

    let ssmDocName = `SHARR-${props.securityStandard}_${props.securityStandardVersion}_${props.controlId}`
    let ssmDocFQFileName = `${props.ssmDocPath}/${props.ssmDocFileName}`
    let ssmDocType = props.ssmDocFileName.substr(props.ssmDocFileName.length - 4).toLowerCase()

    let ssmDocIn = fs.readFileSync(ssmDocFQFileName, 'utf8')

    let ssmDocOut: string = ''
    const re = /^(?<padding>\s+)%%SCRIPT=(?<script>.*)%%/

    for (let line of ssmDocIn.split('\n')) {
        let foundMatch = re.exec(line)
        if (foundMatch && foundMatch.groups && foundMatch.groups.script) {
            let scriptIn = fs.readFileSync(`${scriptPath}/${foundMatch.groups.script}`, 'utf8')
            for (let scriptLine of scriptIn.split('\n')) {
                ssmDocOut += foundMatch.groups.padding + scriptLine + '\n'
            }
        } else {
            ssmDocOut += line + '\n'
        }
    }

    let ssmDocSource = undefined
    if (ssmDocType == 'json') {
        ssmDocSource = JSON.parse(ssmDocOut)
    } else if (ssmDocType == 'yaml') {
        ssmDocSource = yaml.load(ssmDocOut)
    }

    const AutoDoc = new ssm.CfnDocument(this, 'Automation Document', {
        content: ssmDocSource,
        documentType: 'Automation',
        name: ssmDocName
    })
    AutoDoc.cfnOptions.condition = installSsmDoc
  }
}

/*
 * @author AWS Solutions Development
 * @description SSM-based remediation trigger
 * @type {trigger}
 */
export interface ITriggerProps {
    description?: string,
    securityStandard: string;     // ex. AFSBP
    generatorId: string;    // ex. "arn:aws-cn:securityhub:::ruleset/cis-aws-foundations-benchmark/v/1.2.0"
    controlId: string;
    targetArn: string;
}

export class Trigger extends cdk.Construct {
    
  constructor(scope: cdk.Construct, id: string, props: ITriggerProps) {
    super(scope, id);
    let illegalChars = /[\.]/g;

    // Event to Step Function
    // ----------------------
    // Create CWE rule
    // Create custom action

    let description = `Remediate ${props.securityStandard} ${props.controlId}`
    if (props.description) {
        description = props.description
    }

    let workflowStatusFilter = {
        "Status": [ "NEW" ]
    }
    let complianceStatusFilter = {
        "Status": [ "FAILED", "WARNING" ]
    }

    const stateMachine = sfn.StateMachine.fromStateMachineArn(this, 'orchestrator', props.targetArn);

    // Create an IAM role for Events to start the State Machine
    const eventsRole = new Role(this, 'EventsRuleRole', {
      assumedBy: new ServicePrincipal('events.amazonaws.com')
    });

    // Grant the start execution permission to the Events service
    stateMachine.grantStartExecution(eventsRole);

    // Create an event rule to trigger the step function
    const stateMachineTarget: events.IRuleTarget = {
      bind: () => ({
        id: '',
        arn: props.targetArn,
        role: eventsRole
      })
    };

    const enable_auto_remediation_param = new cdk.CfnParameter(this, 'AutoEnable', {
        description: "This will fully enable automated remediation for "+ props.securityStandard + ' ' + props.controlId,
        type: "String",
        allowedValues: ["ENABLED", "DISABLED"],
        default: "DISABLED"
    });

    enable_auto_remediation_param.overrideLogicalId(`${props.securityStandard}${props.controlId.replace(illegalChars, '')}AutoTrigger`)
    
    interface IPattern {
        source: any,
        detailType: any
        detail: any
    }
    let eventPattern: IPattern = {
        source: ["aws.securityhub"],
        detailType: ["Security Hub Findings - Imported"],
        detail: {
            findings: { 
                GeneratorId: [props.generatorId],
                ProductFields: {
                    ControlId: [props.controlId]
                },
                Workflow: workflowStatusFilter,
                Compliance: complianceStatusFilter
            }
        }
    }

    let triggerPattern: events.EventPattern = eventPattern
    
    // Adding an automated even rule for the playbook
    const eventRule_auto = new events.Rule(this, 'AutoEventRule', {
        description: description + ' automatic remediation trigger event rule.',
        ruleName: `${props.securityStandard}_${props.controlId}_AutoTrigger`,
        targets: [stateMachineTarget],
        eventPattern: triggerPattern
    });
    
    const cfnEventRule_auto = eventRule_auto.node.defaultChild as events.CfnRule;
    cfnEventRule_auto.addPropertyOverride('State', enable_auto_remediation_param.valueAsString);
  }
}

export interface IOneTriggerProps {
    description?: string,
    targetArn: string;
    serviceToken: string;
    prereq: cdk.CfnResource[];
}
export class OneTrigger extends cdk.Construct {
// used in place of Trigger. Sends all finding events for which the 
// SHARR custom action is initiated to the Step Function

  constructor(scope: cdk.Construct, id: string, props: IOneTriggerProps) {
    super(scope, id);
    const stack = cdk.Stack.of(this)

    // Event to Step Function
    // ----------------------
    // Create CWE rule
    // Create custom action

    let description = `Remediate with SHARR`
    if (props.description) {
        description = props.description
    }

    let complianceStatusFilter = {
        "Status": [ "FAILED", "WARNING" ]
    }

    const stateMachine = StateMachine.fromStateMachineArn(this, 'orchestrator', props.targetArn);

    // Note: Id is max 20 characters
    const customAction = new cdk.CustomResource(this, 'Custom Action', {
        serviceToken: props.serviceToken,
        resourceType: 'Custom::ActionTarget',
        properties: {
            Name: 'Remediate with SHARR',
            Description: 'Submit the finding to AWS Security Hub Automated Response and Remediation',
            Id: 'SHARRRemediation'
        }
    });
    {
        let child = customAction.node.defaultChild as cdk.CfnCustomResource
        for (var prereq of props.prereq) {
            child.addDependsOn(prereq)
        }
    }

    // Create an IAM role for Events to start the State Machine
    const eventsRole = new Role(this, 'EventsRuleRole', {
    assumedBy: new ServicePrincipal('events.amazonaws.com')
    });

    // Grant the start execution permission to the Events service
    stateMachine.grantStartExecution(eventsRole);

    // Create an event rule to trigger the step function
    const stateMachineTarget: IRuleTarget = {
        bind: () => ({
            id: '',
            arn: props.targetArn,
            role: eventsRole
        })
    };

    let eventPattern: EventPattern = {
        source: ["aws.securityhub"],
        detailType: ["Security Hub Findings - Custom Action"],
        resources: [ customAction.getAttString('Arn') ],
        detail: {
            findings: { 
                Compliance: complianceStatusFilter
            }
        }
    }

    new Rule(this, 'Remediate Custom Action', {
        description: description,
        enabled: true,
        eventPattern: eventPattern,
        ruleName: `Remediate_with_SHARR_CustomAction`,
        targets: [stateMachineTarget]
    })
  }
}

export interface RoleProps {
    readonly solutionId: string;
    readonly ssmDocName: string;
    readonly remediationPolicy: Policy;
    readonly remediationRoleName: string;
}

export class SsmRole extends cdk.Construct {

  constructor(scope: cdk.Construct, id: string, props: RoleProps) {
    super(scope, id);
    const stack = cdk.Stack.of(this)
    const roleStack = MemberRoleStack.of(this)
    const RESOURCE_PREFIX = props.solutionId.replace(/^DEV-/,''); // prefix on every resource name
    const adminRoleName = `${RESOURCE_PREFIX}-SHARR-Orchestrator-Admin`
    const basePolicy = new Policy(this, 'SHARR-Member-Base-Policy')
    const adminAccount = roleStack.node.findChild('AdminAccountParameter').node.findChild('Admin Account Number') as cdk.CfnParameter;

    const ssmParmPerms = new PolicyStatement();
    ssmParmPerms.addActions(
        "ssm:GetParameters",
        "ssm:GetParameter",
        "ssm:PutParameter"
    )
    ssmParmPerms.effect = Effect.ALLOW
    ssmParmPerms.addResources(
        `arn:${stack.partition}:ssm:*:${stack.account}:parameter/Solutions/SO0111/*`
    );
    basePolicy.addStatements(ssmParmPerms)

    // AssumeRole Policy
    let principalPolicyStatement = new PolicyStatement();
    principalPolicyStatement.addActions("sts:AssumeRole");
    principalPolicyStatement.effect = Effect.ALLOW;

    let roleprincipal = new ArnPrincipal(
        'arn:' + stack.partition + ':iam::' + adminAccount.valueAsString +
        ':role/' + adminRoleName
    );

    let principals = new CompositePrincipal(roleprincipal);
    principals.addToPolicy(principalPolicyStatement);

    let serviceprincipal = new ServicePrincipal('ssm.amazonaws.com')
    principals.addPrincipals(serviceprincipal);

    let memberRole = new Role(this, 'MemberAccountRole', {
        assumedBy: principals,
        roleName: props.remediationRoleName
    });

    memberRole.attachInlinePolicy(basePolicy)
    memberRole.attachInlinePolicy(props.remediationPolicy)
    memberRole.applyRemovalPolicy(cdk.RemovalPolicy.RETAIN)

    const memberRoleResource = memberRole.node.findChild('Resource') as CfnRole;

    memberRoleResource.cfnOptions.metadata = {
        cfn_nag: {
            rules_to_suppress: [{
                id: 'W11',
                reason: 'Resource * is required due to the administrative nature of the solution.'
            },{
                id: 'W28',
                reason: 'Static names chosen intentionally to provide integration in cross-account permissions'
            }]
        }
    };
  }
}

export interface RemediationRunbookProps {
    ssmDocName: string;
    ssmDocPath: string;
    ssmDocFileName: string;
    solutionVersion: string;
    solutionDistBucket: string;
    remediationPolicy?: Policy;
    solutionId?: string;
    scriptPath?: string;
}

export class SsmRemediationRunbook extends cdk.Construct {
    
  constructor(scope: cdk.Construct, id: string, props: RemediationRunbookProps) {
    super(scope, id);

    // Add prefix to ssmDocName
    let ssmDocName = `SHARR-${props.ssmDocName}`

    let scriptPath = ''
    if (props.scriptPath == undefined) {
        scriptPath = 'ssmdocs/scripts'
    } else {
        scriptPath = props.scriptPath
    }

    let ssmDocFQFileName = `${props.ssmDocPath}/${props.ssmDocFileName}`
    let ssmDocType = props.ssmDocFileName.substr(props.ssmDocFileName.length - 4).toLowerCase()

    let ssmDocIn = fs.readFileSync(ssmDocFQFileName, 'utf8')

    let ssmDocOut: string = ''
    const re = /^(?<padding>\s+)%%SCRIPT=(?<script>.*)%%/

    for (let line of ssmDocIn.split('\n')) {
        let foundMatch = re.exec(line)
        if (foundMatch && foundMatch.groups && foundMatch.groups.script) {
            let scriptIn = fs.readFileSync(`${scriptPath}/${foundMatch.groups.script}`, 'utf8')
            for (let scriptLine of scriptIn.split('\n')) {
                ssmDocOut += foundMatch.groups.padding + scriptLine + '\n'
            }
        } else {
            ssmDocOut += line + '\n'
        }
    }

    let ssmDocSource = undefined
    if (ssmDocType == 'json') {
        ssmDocSource = JSON.parse(ssmDocOut)
    } else if (ssmDocType == 'yaml') {
        ssmDocSource = yaml.load(ssmDocOut)
    }

    new ssm.CfnDocument(this, 'Automation Document', {
        content: ssmDocSource,
        documentType: 'Automation',
        name: ssmDocName
    })
  }
}
