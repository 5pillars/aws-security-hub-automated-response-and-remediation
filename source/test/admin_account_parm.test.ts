// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
import { SynthUtils } from '@aws-cdk/assert';
import '@aws-cdk/assert/jest';
import { App, Stack } from 'aws-cdk-lib';
import { AdminAccountParm } from '../lib/admin_account_parm-construct';

function createAdminAccountParm(): Stack {
  const app = new App();
  const stack = new Stack(app, 'testStack', {
    stackName: 'testStack',
  });
  new AdminAccountParm(stack, 'roles');
  return stack;
}

test('AdminParm Test Stack', () => {
  expect(SynthUtils.toCloudFormation(createAdminAccountParm())).toMatchSnapshot();
});
