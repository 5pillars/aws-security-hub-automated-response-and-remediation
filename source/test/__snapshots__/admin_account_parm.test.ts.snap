// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdminParm Test Stack 1`] = `
{
  "Parameters": {
    "SecHubAdminAccount": {
      "AllowedPattern": "^\\d{12}$",
      "Description": "Admin account number",
      "Type": "String",
    },
  },
}
`;
